### PR TITLE
JDK-8260571: Add PrintMetaspaceStatistics to print metaspace statistics upon VM exit

### PR DESCRIPTION
**Pull request by tstuefe**
This patch adds a new diagnostic boolean switch `PrintMetaspaceStatistics`. When set, it will cause the VM to print a brief report about metaspace occupancy when exiting. It follows the same logic as `PrintNMTStatistics` and similar switches.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8260571](https://bugs.openjdk.java.net/browse/JDK-8260571): Add PrintMetaspaceStatistics to print metaspace statistics upon VM exit

### Reviewers
 * [Ioi Lam](https://openjdk.java.net/census#iklam) (@iklam - **Reviewer**) ⚠️ Review applies to 9959026778c52cbd4c3d9627781d9b88914b5ab8

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2283/head:pull/2283`
`$ git checkout pull/2283`

### CODE REVIEW

**Comment by iklam** [comment]
I am wondering if we should use UL to avoid adding new command-line options.  How about this in `print_statistics()` in java.c? This way you can also distinguish the output with other "legacy" logs such as `PrintSystemDictionaryAtExit` (because `MetaspaceUtils::print_basic_report` doesn't indicate where its output starts and ends).

```
LogStreamHandle(Info, metaspace) metaspace_log;
if (metaspace_log.is_enabled()) {
  MetaspaceUtils::print_basic_report(&metaspace_log, 0);
}
```

Eventually we should move `PrintSystemDictionaryAtExit`, etc, to UL as well. We may need to add a new `atexit` tag for flexibility, so you can say

```
java -Xlog:systemdictionary+atexit -Xlog:metaspace+atexit:file=metsapce.log
```

**Comment by tstuefe** [comment]
> I am wondering if we should use UL to avoid adding new command-line options. How about this in `print_statistics()` in java.c? This way you can also distinguish the output with other "legacy" logs such as `PrintSystemDictionaryAtExit` (because `MetaspaceUtils::print_basic_report` doesn't indicate where its output starts and ends).
> 
> ```
> LogStreamHandle(Info, metaspace) metaspace_log;
> if (metaspace_log.is_enabled()) {
>   MetaspaceUtils::print_basic_report(&metaspace_log, 0);
> }
> ```
> 
> Eventually we should move `PrintSystemDictionaryAtExit`, etc, to UL as well. We may need to add a new `atexit` tag for flexibility, so you can say
> 
> ```
> java -Xlog:systemdictionary+atexit -Xlog:metaspace+atexit:file=metsapce.log
> ```

I rather would leave the translation of the "XXAtExit" flags to UL - if it even happens in that way - for anther patch instead of mixing those two unrelated issues together. Easier to backport, for one thing.

Note that I am also not convinced that is the right way to do. I'm not completely against it but tbh I find the way UL handles tags very strange, with the tags being a jumble of component names, phase descriptors and what feels like just random words. It is certainly a uncommon system and has a number of drawbacks. So I wonder whether it would make sense to have a discussion before further going down this road.

File: src/hotspot/share/runtime/globals.hpp [hunk 1]
```diff
@@ -1484,6 +1484,9 @@ const intx ObjectAlignmentInBytes = 8;
   product(ccstr, MetaspaceReclaimPolicy, "balanced",                        \
           "options: balanced, aggressive, none")                            \
                                                                             \
+  product(bool, PrintMetaspaceStatistics, false, DIAGNOSTIC,                \
```
Comment (coleenp) [inline src/hotspot/share/runtime/globals.hpp:1487]: Can you call this PrintMetaspaceStatisticsAtExit to be like the other ones?

**Review by coleenp — COMMENTED** [review]
Looks good to me.

**Comment by coleenp** [comment]
I agree with Thomas.  I think there's a way to make UL not print the decorations with these Print<Stuff>AtExit, but otherwise the decorations are noisy.  We purposefully didn't not translate the options that print a lot of output to UL because the decorations and the timestamps aren't interesting. eg. PrintSystemDictionaryAtExit or PrintInterpreter.  If someone wants to change this so that these options suppress decorations and translate to UL, we can change metspace printing at exit also.  As a diagnostic flag, there isn't any promise of it's longevity in the code. ie. we can change it if we want.

**Comment by iklam** [comment]
> I agree with Thomas. I think there's a way to make UL not print the decorations with these PrintAtExit, but otherwise the decorations are noisy. We purposefully didn't not translate the options that print a lot of output to UL because the decorations and the timestamps aren't interesting. eg. PrintSystemDictionaryAtExit or PrintInterpreter. If someone wants to change this so that these options suppress decorations and translate to UL, we can change metspace printing at exit also. As a diagnostic flag, there isn't any promise of it's longevity in the code. ie. we can change it if we want.

OK, I withdraw my objection.

**Review by iklam — APPROVED** [review]

File: src/hotspot/share/runtime/globals.hpp (hunk 1 above)
Comment (tstuefe) [reply src/hotspot/share/runtime/globals.hpp:1487]: Sure!

**Comment by tstuefe** [comment]
Coleen, Ioi, thank you!
I'll rename the switch to `PrintMetaspaceStatisticsAtExit` as Coleen requested and then push.

> Looks good to me.

Thanks, Coleen!

**Comment by tstuefe** [comment]
> > I agree with Thomas. I think there's a way to make UL not print the decorations with these PrintAtExit, but otherwise the decorations are noisy. We purposefully didn't not translate the options that print a lot of output to UL because the decorations and the timestamps aren't interesting. eg. PrintSystemDictionaryAtExit or PrintInterpreter. If someone wants to change this so that these options suppress decorations and translate to UL, we can change metspace printing at exit also. As a diagnostic flag, there isn't any promise of it's longevity in the code. ie. we can change it if we want.
> 
> OK, I withdraw my objection.

Thanks!

**Pull request closed without merging**